File: benchmark/config_validation.dart
import 'support/comparison.dart';

// Validate service manifests with nested routes and optional metadata.
// dart compile exe benchmark/config_validation.dart -o .dart_tool/config-baseline.exe
// .dart_tool/config-baseline.exe [batches] [samples]
const _source = r'''
class ConfigError implements Exception {
  ConfigError(this.path, this.reason);
  final String path;
  final String reason;
}

class Route {
  Route(this.path, this.target);
  final String path;
  final String target;

  int get score => path.length * 11 + target.length * 7;
}

class ServiceConfig {
  ServiceConfig(this.name, this.host, this.port, {
    required this.enabled,
    required this.retries,
    required this.owner,
    required this.tags,
    required this.routes,
  });
  final String name;
  final String host;
  final int port;
  final bool enabled;
  final int retries;
  final String owner;
  final List<String> tags;
  final List<Route> routes;

  int get score {
    var total = name.length * 3 + host.length + port + retries * 5 +
        owner.length + (enabled ? 17 : 0);
    for (final tag in tags) {
      total += tag.length * 13;
    }
    for (final route in routes) {
      total += route.score;
    }
    return total;
  }
}

Map<String, Object?> requireMap(Object? value, String path) {
  if (value is! Map<String, Object?>) throw ConfigError(path, 'type');
  return value;
}

String requireString(Map<String, Object?> values, String key, String path) {
  final value = values[key];
  if (value is! String || value.isEmpty) throw ConfigError(path, 'string');
  return value;
}

int requireInt(Map<String, Object?> values, String key, String path) {
  final value = values[key];
  if (value is! int) throw ConfigError(path, 'integer');
  return value;
}

Route parseRoute(Object? raw, String path) {
  final values = requireMap(raw, path);
  final routePath = requireString(values, 'path', '$path.path');
  if (!routePath.startsWith('/')) throw ConfigError('$path.path', 'absolute');
  final target = requireString(values, 'target', '$path.target');
  return Route(routePath, target);
}

ServiceConfig validate(Object? raw) {
  final values = requireMap(raw, 'service');
  final name = requireString(values, 'name', 'service.name');
  final host = requireString(values, 'host', 'service.host');
  final port = requireInt(values, 'port', 'service.port');
  if (port < 1 || port > 65535) throw ConfigError('service.port', 'range');

  final enabledValue = values['enabled'];
  if (enabledValue != null && enabledValue is! bool) {
    throw ConfigError('service.enabled', 'boolean');
  }
  final enabled = enabledValue == null ? true : enabledValue as bool;

  final retriesValue = values['retries'];
  if (retriesValue != null && retriesValue is! int) {
    throw ConfigError('service.retries', 'integer');
  }
  final retries = retriesValue == null ? 2 : retriesValue as int;
  if (retries < 0 || retries > 5) throw ConfigError('service.retries', 'range');

  final routesValue = values['routes'];
  if (routesValue is! List) throw ConfigError('service.routes', 'list');
  final routes = <Route>[];
  for (var i = 0; i < routesValue.length; i++) {
    routes.add(parseRoute(routesValue[i], 'service.routes[$i]'));
  }

  final metadataValue = values['metadata'];
  var owner = 'unknown';
  final tags = <String>[];
  if (metadataValue != null) {
    final metadata = requireMap(metadataValue, 'service.metadata');
    final ownerValue = metadata['owner'];
    if (ownerValue != null) {
      owner = requireString(metadata, 'owner', 'service.metadata.owner');
    }
    final tagsValue = metadata['tags'];
    if (tagsValue != null) {
      if (tagsValue is! List) throw ConfigError('service.metadata.tags', 'list');
      for (var i = 0; i < tagsValue.length; i++) {
        final tag = tagsValue[i];
        if (tag is! String || tag.isEmpty) {
          throw ConfigError('service.metadata.tags[$i]', 'string');
        }
        tags.add(tag);
      }
    }
  }
  return ServiceConfig(name, host, port,
      enabled: enabled, retries: retries, owner: owner, tags: tags, routes: routes);
}

int main(int batches) {
  final configs = <Map<String, Object?>>[
    {'name': 'web', 'host': 'web.internal', 'port': 8080, 'enabled': true,
      'retries': 3, 'routes': [
        {'path': '/api', 'target': 'api'},
        {'path': '/health', 'target': 'status'},
      ], 'metadata': {'owner': 'core', 'tags': ['public', 'api']}},
    {'name': 'worker', 'host': 'queue.internal', 'port': 9001,
      'enabled': false, 'routes': []},
    {'name': 'metrics', 'host': 'stats.internal', 'port': 9100,
      'routes': [{'path': '/metrics', 'target': 'prometheus'}],
      'metadata': {'owner': 'ops', 'tags': []}},
    {'name': 'gateway', 'host': 'edge.example', 'port': 443,
      'retries': null, 'routes': [
        {'path': '/', 'target': 'web'},
        {'path': '/login', 'target': 'auth'},
        {'path': '/static', 'target': 'assets'},
      ]},
    {'name': 'docs', 'host': 'docs.internal', 'port': 8081,
      'routes': [{'path': '/guide', 'target': 'pages'}],
      'metadata': {'owner': null, 'tags': ['web']}},
    {'name': 'sync', 'host': 'sync.internal', 'port': 7001,
      'routes': [], 'metadata': {'owner': 'data', 'tags': null}},
    {'name': 'notifications', 'host': 'mail.internal', 'port': 2525,
      'enabled': null, 'retries': 1, 'routes': [
        {'path': '/send', 'target': 'mailer'},
        {'path': '/status', 'target': 'queue'},
      ]},
    {'name': 'audit', 'host': 'audit.internal', 'port': 4111,
      'routes': [{'path': '/events', 'target': 'log'}]},
    {'name': 'broken-port', 'host': 'bad.internal', 'port': 70000,
      'routes': []},
    {'name': 'broken-route', 'host': 'bad.internal', 'port': 8080,
      'routes': [{'path': 'health', 'target': 'status'}]},
  ];

  var checksum = 0;
  var valid = 0;
  var invalid = 0;
  for (var batch = 0; batch < batches; batch++) {
    for (final raw in configs) {
      try {
        checksum += validate(raw).score;
        valid++;
      } on ConfigError catch (error) {
        checksum += error.path.length * 13 + error.reason.length * 7;
        invalid++;
      }
    }
  }
  return checksum + valid * 1000003 + invalid * 1009;
}
''';

void main(List<String> args) => runComparison(
  args,
  name: 'config_validation',
  source: _source,
  parameter: 'batches',
  unit: 'batch',
  iterations: 500,
  warmupIterations: 5,
);

File: benchmark/config_validation.py
from support.comparison import run_comparison


class ConfigError(Exception):
    def __init__(self, path: str, reason: str) -> None:
        self.path = path
        self.reason = reason


class Route:
    def __init__(self, path: str, target: str) -> None:
        self.path = path
        self.target = target

    @property
    def score(self) -> int:
        return len(self.path) * 11 + len(self.target) * 7


class ServiceConfig:
    def __init__(self, name: str, host: str, port: int, *, enabled: bool,
                 retries: int, owner: str, tags: list[str], routes: list[Route]) -> None:
        self.name = name
        self.host = host
        self.port = port
        self.enabled = enabled
        self.retries = retries
        self.owner = owner
        self.tags = tags
        self.routes = routes

    @property
    def score(self) -> int:
        total = (len(self.name) * 3 + len(self.host) + self.port +
                 self.retries * 5 + len(self.owner) + (17 if self.enabled else 0))
        for tag in self.tags:
            total += len(tag) * 13
        for route in self.routes:
            total += route.score
        return total


def require_map(value: object, path: str) -> dict[str, object]:
    if type(value) is not dict:
        raise ConfigError(path, 'type')
    return value


def require_string(values: dict[str, object], key: str, path: str) -> str:
    value = values.get(key)
    if type(value) is not str or not value:
        raise ConfigError(path, 'string')
    return value


def require_int(values: dict[str, object], key: str, path: str) -> int:
    value = values.get(key)
    if type(value) is not int:
        raise ConfigError(path, 'integer')
    return value


def parse_route(raw: object, path: str) -> Route:
    values = require_map(raw, path)
    route_path = require_string(values, 'path', f'{path}.path')
    if not route_path.startswith('/'):
        raise ConfigError(f'{path}.path', 'absolute')
    target = require_string(values, 'target', f'{path}.target')
    return Route(route_path, target)


def validate(raw: object) -> ServiceConfig:
    values = require_map(raw, 'service')
    name = require_string(values, 'name', 'service.name')
    host = require_string(values, 'host', 'service.host')
    port = require_int(values, 'port', 'service.port')
    if port < 1 or port > 65535:
        raise ConfigError('service.port', 'range')

    enabled_value = values.get('enabled')
    if enabled_value is not None and type(enabled_value) is not bool:
        raise ConfigError('service.enabled', 'boolean')
    enabled = True if enabled_value is None else enabled_value

    retries_value = values.get('retries')
    if retries_value is not None and type(retries_value) is not int:
        raise ConfigError('service.retries', 'integer')
    retries = 2 if retries_value is None else retries_value
    if retries < 0 or retries > 5:
        raise ConfigError('service.retries', 'range')

    routes_value = values.get('routes')
    if type(routes_value) is not list:
        raise ConfigError('service.routes', 'list')
    routes = []
    for i in range(len(routes_value)):
        routes.append(parse_route(routes_value[i], f'service.routes[{i}]'))

    metadata_value = values.get('metadata')
    owner = 'unknown'
    tags = []
    if metadata_value is not None:
        metadata = require_map(metadata_value, 'service.metadata')
        owner_value = metadata.get('owner')
        if owner_value is not None:
            owner = require_string(metadata, 'owner', 'service.metadata.owner')
        tags_value = metadata.get('tags')
        if tags_value is not None:
            if type(tags_value) is not list:
                raise ConfigError('service.metadata.tags', 'list')
            for i in range(len(tags_value)):
                tag = tags_value[i]
                if type(tag) is not str or not tag:
                    raise ConfigError(f'service.metadata.tags[{i}]', 'string')
                tags.append(tag)
    return ServiceConfig(name, host, port, enabled=enabled, retries=retries,
                         owner=owner, tags=tags, routes=routes)


def main(batches: int) -> int:
    configs = [
        {'name': 'web', 'host': 'web.internal', 'port': 8080, 'enabled': True,
         'retries': 3, 'routes': [
             {'path': '/api', 'target': 'api'},
             {'path': '/health', 'target': 'status'},
         ], 'metadata': {'owner': 'core', 'tags': ['public', 'api']}},
        {'name': 'worker', 'host': 'queue.internal', 'port': 9001,
         'enabled': False, 'routes': []},
        {'name': 'metrics', 'host': 'stats.internal', 'port': 9100,
         'routes': [{'path': '/metrics', 'target': 'prometheus'}],
         'metadata': {'owner': 'ops', 'tags': []}},
        {'name': 'gateway', 'host': 'edge.example', 'port': 443,
         'retries': None, 'routes': [
             {'path': '/', 'target': 'web'},
             {'path': '/login', 'target': 'auth'},
             {'path': '/static', 'target': 'assets'},
         ]},
        {'name': 'docs', 'host': 'docs.internal', 'port': 8081,
         'routes': [{'path': '/guide', 'target': 'pages'}],
         'metadata': {'owner': None, 'tags': ['web']}},
        {'name': 'sync', 'host': 'sync.internal', 'port': 7001,
         'routes': [], 'metadata': {'owner': 'data', 'tags': None}},
        {'name': 'notifications', 'host': 'mail.internal', 'port': 2525,
         'enabled': None, 'retries': 1, 'routes': [
             {'path': '/send', 'target': 'mailer'},
             {'path': '/status', 'target': 'queue'},
         ]},
        {'name': 'audit', 'host': 'audit.internal', 'port': 4111,
         'routes': [{'path': '/events', 'target': 'log'}]},
        {'name': 'broken-port', 'host': 'bad.internal', 'port': 70000,
         'routes': []},
        {'name': 'broken-route', 'host': 'bad.internal', 'port': 8080,
         'routes': [{'path': 'health', 'target': 'status'}]},
    ]

    checksum = 0
    valid = 0
    invalid = 0
    for _ in range(batches):
        for raw in configs:
            try:
                checksum += validate(raw).score
                valid += 1
            except ConfigError as error:
                checksum += len(error.path) * 13 + len(error.reason) * 7
                invalid += 1
    return checksum + valid * 1000003 + invalid * 1009


if __name__ == '__main__':
    run_comparison('config_validation', main, unit='batch', iterations=500,
                   warmup_iterations=5)
